### PR TITLE
Add `SASL-SCRAM-256` & `SASL-SCRAM-512` authentication support

### PR DESCRIPTION
**Pull request by dilanSachi**
## Purpose
$subject

Fixes https://github.com/ballerina-platform/ballerina-library/issues/6766
Fixes https://github.com/ballerina-platform/ballerina-library/issues/6763

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [x] Checked native-image compatibility

### CODE REVIEW

File: ballerina/tests/producer_client_tests.bal [hunk 1]
```diff
@@ -304,6 +304,62 @@ function saslProducerTest() returns error? {
...
+
+    ProducerConfiguration producerConfigs = {
+        clientId: "test-producer-06",
+        auth: authScram256Config,
+        securityProtocol: PROTOCOL_SASL_PLAINTEXT
+    };
+
+    Producer kafkaProducer = check new (SASL_URL, producerConfigs);
+    Error? result = kafkaProducer->send({topic: topic, value: TEST_MESSAGE.toBytes() });
+    test:assertFalse(result is error, result is error ? result.toString() : result.toString());
```
Comment (ayeshLK) [inline ballerina/tests/producer_client_tests.bal:320]: Any reason to follow this structure ? We could have done this IMO
```suggestion
    check kafkaProducer->send({topic: topic, value: TEST_MESSAGE.toBytes() });
```

File: native/src/main/java/io/ballerina/stdlib/kafka/utils/KafkaUtils.java [hunk 2]
```diff
@@ -372,6 +374,18 @@ private static void processSaslProperties(BMap<BString, Object> configurations,
             addStringParamIfPresent(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, authenticationConfig, properties,
                                     KafkaConstants.SECURITY_PROTOCOL_CONFIG);
             properties.put(SaslConfigs.SASL_JAAS_CONFIG, jaasConfigValue);
+        } else if (KafkaConstants.SASL_SCRAM_SHA_256.equals(mechanism) ||
+                KafkaConstants.SASL_SCRAM_SHA_512.equals(mechanism)) {
+            String username = authenticationConfig.getStringValue(KafkaConstants.USERNAME).getValue();
+            String password = authenticationConfig.getStringValue(KafkaConstants.PASSWORD).getValue();
+            String jaasConfigValue =
+                    "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"" + username +
+                            "\" password=\"" + password + "\";";
```
Comment (ayeshLK) [inline native/src/main/java/io/ballerina/stdlib/kafka/utils/KafkaUtils.java:383]: Can't we do this ?
```suggestion
            String jaasConfigValue = String.format("org.apache.kafka.common.security.scram.ScramLoginModule required username=\"%s\" password=\"%s\";", username, password);
```

File: ballerina/tests/producer_client_tests.bal [hunk 3]
```diff
@@ -304,6 +304,62 @@ function saslProducerTest() returns error? {
...
+
+    ConsumerConfiguration consumerConfiguration = {
+        topics: [topic],
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "sasl-scram-512-producer-test-group",
+        clientId: "test-consumer-49"
+    };
+    Consumer consumer = check new (DEFAULT_URL, consumerConfiguration);
+    BytesConsumerRecord[] consumerRecords = check consumer->poll(5);
+    test:assertEquals(consumerRecords.length(), 1, "Expected: 1. Received: " + consumerRecords.length().toString());
```
Comment (ayeshLK) [inline ballerina/tests/producer_client_tests.bal:359]:
```suggestion
    test:assertEquals(consumerRecords.length(), 1, string `Expected: 1. Received: ${consumerRecords.length()}`);
```

File: ballerina/tests/producer_client_tests.bal [hunk 4]
```diff
@@ -304,6 +304,62 @@ function saslProducerTest() returns error? {
...
+
+    ConsumerConfiguration consumerConfiguration = {
+        topics: [topic],
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "sasl-scram-256-producer-test-group",
+        clientId: "test-consumer-49"
+    };
+    Consumer consumer = check new (DEFAULT_URL, consumerConfiguration);
+    BytesConsumerRecord[] consumerRecords = check consumer->poll(5);
+    test:assertEquals(consumerRecords.length(), 1, "Expected: 1. Received: " + consumerRecords.length().toString());
```
Comment (ayeshLK) [inline ballerina/tests/producer_client_tests.bal:331]:
```suggestion
    test:assertEquals(consumerRecords.length(), 1, string `Expected: 1. Received: ${consumerRecords.length()}`);
```

File: ballerina/tests/consumer_client_tests.bal [hunk 5]
```diff
@@ -995,6 +1011,48 @@ function saslConsumerTest() returns error? {
...
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "consumer-sasl-scram-512-test-group",
+        clientId: "test-consumer-42",
+        autoCommit: false,
+        auth: authScram512Config,
+        securityProtocol: PROTOCOL_SASL_PLAINTEXT
+    };
+    Consumer consumer = check new(SASL_URL, consumerConfiguration);
+    BytesConsumerRecord[] consumerRecords = check consumer->poll(5);
+    test:assertEquals(consumerRecords.length(), 1, "Expected: 1. Received: " + consumerRecords.length().toString());
```
Comment (ayeshLK) [inline ballerina/tests/consumer_client_tests.bal:1052]:
```suggestion
    test:assertEquals(consumerRecords.length(), 1, string `Expected: 1. Received: ${consumerRecords.length()}`);
```

File: ballerina/tests/consumer_client_tests.bal [hunk 6]
```diff
@@ -995,6 +1011,48 @@ function saslConsumerTest() returns error? {
...
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "consumer-sasl-scram-256-test-group",
+        clientId: "test-consumer-42",
+        autoCommit: false,
+        auth: authScram256Config,
+        securityProtocol: PROTOCOL_SASL_PLAINTEXT
+    };
+    Consumer consumer = check new(SASL_URL, consumerConfiguration);
+    BytesConsumerRecord[] consumerRecords = check consumer->poll(5);
+    test:assertEquals(consumerRecords.length(), 1, "Expected: 1. Received: " + consumerRecords.length().toString());
```
Comment (ayeshLK) [inline ballerina/tests/consumer_client_tests.bal:1031]:
```suggestion
    test:assertEquals(consumerRecords.length(), 1, string `Expected: 1. Received: ${consumerRecords.length()}`);
```

File: native/src/main/java/io/ballerina/stdlib/kafka/utils/KafkaUtils.java (hunk 2 above)
Comment (dilanSachi) [reply native/src/main/java/io/ballerina/stdlib/kafka/utils/KafkaUtils.java]: Fixed

File: ballerina/tests/producer_client_tests.bal (hunk 1 above)
Comment (dilanSachi) [reply ballerina/tests/producer_client_tests.bal:320]: Will attend to these test case refactors separately

File: ballerina/tests/producer_client_tests.bal [hunk 7]
```diff
@@ -304,6 +304,62 @@ function saslProducerTest() returns error? {
...
+
+    ProducerConfiguration producerConfigs = {
+        clientId: "test-producer-06",
+        auth: authScram512Config,
+        securityProtocol: PROTOCOL_SASL_PLAINTEXT
+    };
+
+    Producer kafkaProducer = check new (SASL_URL, producerConfigs);
+    Error? result = kafkaProducer->send({topic: topic, value: TEST_MESSAGE.toBytes() });
+    test:assertFalse(result is error, result is error ? result.toString() : result.toString());
```
[the remark this answers is not in the document]
Comment (dilanSachi) [reply ballerina/tests/producer_client_tests.bal:348]: Will attend to these test case refactors separately

File: ballerina/tests/producer_client_tests.bal (hunk 3 above)
Comment (dilanSachi) [reply ballerina/tests/producer_client_tests.bal:359]: Will attend to these test case refactors separately

File: ballerina/tests/producer_client_tests.bal (hunk 4 above)
Comment (dilanSachi) [reply ballerina/tests/producer_client_tests.bal:331]: Will attend to these test case refactors separately

**Review by shafreenAnfar — APPROVED** [review]

**Review by Nuvindu — APPROVED** [review]

**Review by Nuvindu — APPROVED** [review]

**Pull request merged**